scripts/dts/gen_defines.py

```diff
@@ -969,181 +292,211 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
-
-    def write_clock_other(ident, clock_i, clock, primary_cdf):
-        # Write clock macros for a secondary identifier 'ident'
-
-        prim_controller, prim_data, prim_freq = primary_cdf
-        if prim_controller is not None:
-            # FIXME this is replicating previous behavior that didn't
-            # generate an indexed controller bug-for-bug. There's a
-            # missing _{clock_i} at the end of the f-string.
-            out(f"{ident}_CLOCK_CONTROLLER", prim_controller)
-        for name_i, name in enumerate(clock.data.keys()):
-            out(f"{ident}_CLOCK_{str2ident(name)}_{clock_i}", prim_data[name_i])
-            if clock_i == 0:
-                out(f"{ident}_CLOCK_{str2ident(name)}", prim_data[name_i])
-        if prim_freq is not None:
-            # FIXME this is also a bug-for-bug match with the previous
-            # implementation.
-            out(f"{ident}_CLOCKS_CLOCK_FREQUENCY", prim_freq)
-
-    out_comment("Clock gate macros from the 'clocks' property",
-                blank_before=False)
-    for_each_ident(node, write_clocks_for_ident)
+
+def controller_and_data_macros(entry, prefix, suffix):
+    # Helper procedure shared by phandle_array_macros() and
+    # write_spi_cs_gpio(). The comments before each macro definition
+    # are for the phandle_array_macros() case, since it's the primary
+    # user.
+
+    ret = {}
+    ctlr = quote_str(entry.controller.label)
+    data = entry.data
+
+    # DT_N_<node-id>_P_<prop-id>_IDX_<i>_BY_CTLR
+    ret[f"{prefix}{suffix}_BY_CTLR"] = list2init(
+        [ctlr] + [str(value) for value in data.values()])
+    # DT_N_<node-id>_P_<prop-id>_IDX_<i>_CELL_CTLR
+    ret[f"{prefix}{suffix}_CELL_CTLR"] = ctlr
+    # DT_N_<node-id>_P_<prop-id>_IDX_<i>_CELL_<CELL>
+    for cell, val in data.items():
+        ret[f"{prefix}{suffix}_CELL_{str2ident(cell)}"] = val
+
+    if not entry.name:
+        return ret
+
+    name = str2ident(entry.name)
+    # DT_N_<node-id>_P_<prop-id>_IDX_<i>_NAME
+    ret[f"{prefix}{suffix}_NAME"] = quote_str(entry.name)
+    # DT_N_<node-id>_P_<prop-id>_NAME_<NAME>_BY_CTLR
+    ret[f"{prefix}_NAME_{name}_BY_CTLR"] = f"DT_{prefix}{suffix}_BY_CTLR"
+    # DT_N_<node-id>_P_<prop-id>_NAME_<NAME>_CELL_CTLR
+    ret[f"{prefix}_NAME_{name}_CELL_CTLR"] = f"DT_{prefix}{suffix}_CELL_CTLR"
+    # DT_N_<node-id>_P_<prop-id>_NAME_<NAME>_CELL_<CELL>
+    for cell, val in data.items():
+        cell_ident = str2ident(cell)
+        ret[f"{prefix}_NAME_{name}_CELL_{cell_ident}"] = \
+            f"DT_{prefix}{suffix}_CELL_{cell_ident}"
+
+    return ret
 
 
 def str2ident(s):
     # Converts 's' to a form suitable for (part of) an identifier
 
-    return s.replace("-", "_") \
-            .replace(",", "_") \
-            .replace("@", "_") \
-            .replace("/", "_") \
-            .replace(".", "_") \
-            .replace("+", "PLUS") \
-            .upper()
+    return re.sub('[-,.@/+]', '_', s.lower())
```

Comment:
"+" is turned into "PLUS" to give nicer output for stuff like `compatible = "arm,cortex-m0+"`.

Reply:
> "+" is turned into "PLUS" to give nicer output for stuff like `compatible = "arm,cortex-m0+"`.

Thanks for explaining the usage. I don't think the benefit to the output is worth the extra complication in the documentation, though. We would have to:

- say "special characters are replaced with underscores, *except* for plus, which is replaced with PLUS", and 
- replace "lowercase-and-underscores" with something like "lowercase-and-underscores-except-plus" everywhere in devicetree.h's doxygen strings

Reply:
Up to you. I was just copying the old logic.